refactor: rename Archiver to ArchiveStore

nazarhussain opened the pull request:
**Motivation**

This is part of the series of the PRs. The idea is the **Archives**, **Historical State** and **Archiver** have similar scope in the system, these all deals with the data which is to be archived or archived already. So we converge those all into one module called **ArchiveStore** , it will be responsible to storing and returning archives when needed. 

**Description**

- Rename `archiver` module to `archiveStore`
- Rename `Archiver` class to `ArchiveStore`

**Steps to test or reproduce**

- Run all tests

nazarhussain committed:
chore: rename Archiver to ArchiveStore (#7518)

**Motivation**

This is part of the series of the PRs. The idea is the **Archives**,
**Historical State** and **Archiver** have similar scope in the system,
these all deals with the data which is to be archived or archived
already. So we converge those all into one module called
**ArchiveStore** , it will be responsible to storing and returning
archives when needed.

**Description**

- Rename `archiver` module to `archiveStore`
- Rename `Archiver` class to `ArchiveStore`

**Steps to test or reproduce**

Run all tests 


**NOTE: The upcoming PR will merge the `historicalState` module to the
`archiveStore`**

nazarhussain committed:
fix: import path for modules

matthewkeil reviewed (dismissed):
LGTM!!! 🎸

nflaig reviewed (commented):
is this PR intended to target unstable branch?

nazarhussain commented:
> is this PR intended to target unstable branch?

Yes, but not merging it now. There are few more PRs going to be added into feature branch before.

nflaig commented on packages/beacon-node/src/chain/archiveStore/constants.ts:
```diff
@@ -0,0 +1,5 @@
+
+export const DEFAULT_ARCHIVE_MODE = ArchiveMode.Frequency;
+
+export const PROCESS_FINALIZED_CHECKPOINT_QUEUE_LEN = 256;
```
this variable is already long, might as well not abbreviate length
```suggestion
export const PROCESS_FINALIZED_CHECKPOINT_QUEUE_LENGTH = 256;
```

nflaig commented on packages/cli/src/options/beaconNodeOptions/chain.ts:
```diff
@@ -231,11 +231,11 @@ Will double processing times. Use only for debugging purposes.",
   },
 
-  "chain.stateArchiveMode": {
+  "chain.archiveMode": {
```
curious about the rationale for renaming, so we are combining block and state archive now?

nazarhussain replied:
We want `ArchiveStore` to have only one `mode` and handles any types of archives based on that mode. So why stripped the `state` from there.

nazarhussain committed:
refactor: extract extra operations from archive store to utils (#7592)

**Motivation**

Make the archive store as slim possible with only publicly useful
interface.

**Description**

- Fix mis-spelled name `ArchvierStore` to `ArchiveStore`. 
- Move `updateBackFillRange` to a pure util function

**Steps to test or reproduce**

Run all tests

matthewkeil reviewed (changes requested):
Just noticed other PR will merge into this. Dismissing my approval for final code review

nazarhussain committed:
refactor: encapsulate historical state inside archive store (#7593)

**Motivation**

As the name suggests the `ArchiveStore` keeps track of any archives
hence it should be also responsible for historical state. This will
helps to encapsulate this logic and introduce differential state
archives without breaking any interface.

**Description**

- Move historicalStateRegen to archiveStore
- Add `init` method in archiveStore which takes care of a few scattered
history related actions


**Steps to test or reproduce**

- Run all tests

---------

Co-authored-by: twoeths <10568965+twoeths@users.noreply.github.com>

nazarhussain committed:
Rename postfix of a constant

nazarhussain committed:
Merge branch 'unstable' into feature/archive-store-2

matthewkeil reviewed (approved):
LGTM!! Just waiting for CI to go green 🚀

nflaig commented on packages/beacon-node/src/chain/archiveStore/archiveStore.ts:
as noted in my previous comment https://github.com/ChainSafe/lodestar/pull/7437#issuecomment-2675398318 I really dislike having a diff as in this PR, it not only makes it harder to review but also makes it hard to inspect file changes later on as git blame doesn't work very well with such changes

this is how the actual diff should have looked like if the file renaming was done separately
```diff
diff --git a/packages/beacon-node/src/chain/archiveStore/archiveStore.ts b/packages/beacon-node/src/chain/archiveStore/archiveStore.ts
index 40a011082c..d4e1743a92 100644
--- a/packages/beacon-node/src/chain/archiveStore/archiveStore.ts
+++ b/packages/beacon-node/src/chain/archiveStore/archiveStore.ts
@@ -1,58 +1,82 @@
 import {CheckpointWithHex} from "@lodestar/fork-choice";
-import {Logger} from "@lodestar/utils";
+import {LoggerNode} from "@lodestar/logger/node";
+import {Checkpoint} from "@lodestar/types/phase0";
+import {callFnWhenAwait} from "@lodestar/utils";
 import {IBeaconDb} from "../../db/index.js";
 import {Metrics} from "../../metrics/metrics.js";
 import {JobItemQueue} from "../../util/queue/index.js";
 import {ChainEvent} from "../emitter.js";
 import {IBeaconChain} from "../interface.js";
-import {archiveBlocks} from "./archiveBlocks.js";
-import {ArchiverOpts, StateArchiveMode, StateArchiveStrategy} from "./interface.js";
-import {pruneHistory} from "./pruneHistory.js";
+import {PROCESS_FINALIZED_CHECKPOINT_QUEUE_LENGTH} from "./constants.js";
+import {HistoricalStateRegen} from "./historicalState/historicalStateRegen.js";
+import {ArchiveMode, ArchiveStoreOpts, StateArchiveStrategy} from "./interface.js";
 import {FrequencyStateArchiveStrategy} from "./strategies/frequencyStateArchiveStrategy.js";
+import {archiveBlocks} from "./utils/archiveBlocks.js";
+import {pruneHistory} from "./utils/pruneHistory.js";
+import {updateBackfillRange} from "./utils/updateBackfillRange.js";
 
-export const DEFAULT_STATE_ARCHIVE_MODE = StateArchiveMode.Frequency;
+type ArchiveStoreModules = {
+  chain: IBeaconChain;
+  db: IBeaconDb;
+  logger: LoggerNode;
+  metrics: Metrics | null;
+};
 
-export const PROCESS_FINALIZED_CHECKPOINT_QUEUE_LEN = 256;
+type ArchiveStoreInitOpts = ArchiveStoreOpts & {dbName: string; anchorState: {finalizedCheckpoint: Checkpoint}};
 
 /**
  * Used for running tasks that depends on some events or are executed
  * periodically.
  */
-export class Archiver {
-  private stateArchiveMode: StateArchiveMode;
+export class ArchiveStore {
+  private archiveMode: ArchiveMode;
   private jobQueue: JobItemQueue<[CheckpointWithHex], void>;
 
   private prevFinalized: CheckpointWithHex;
-  private readonly statesArchiverStrategy: StateArchiveStrategy;
   private archiveBlobEpochs?: number;
-
-  constructor(
-    private readonly db: IBeaconDb,
-    private readonly chain: IBeaconChain,
-    private readonly logger: Logger,
-    signal: AbortSignal,
-    private readonly opts: ArchiverOpts,
-    private readonly metrics?: Metrics | null
-  ) {
-    if (opts.stateArchiveMode === StateArchiveMode.Frequency) {
-      this.statesArchiverStrategy = new FrequencyStateArchiveStrategy(chain.regen, db, logger, opts, chain.bufferPool);
-    } else {
-      throw new Error(`State archive strategy "${opts.stateArchiveMode}" currently not supported.`);
-    }
-
-    this.stateArchiveMode = opts.stateArchiveMode;
+  private readonly statesArchiverStrategy: StateArchiveStrategy;
+  private readonly chain: IBeaconChain;
+  private readonly db: IBeaconDb;
+  private readonly logger: LoggerNode;
+  private readonly metrics: Metrics | null;
+  private readonly opts: ArchiveStoreInitOpts;
+  private readonly signal: AbortSignal;
+
+  private historicalStateRegen?: HistoricalStateRegen;
+
+  constructor(modules: ArchiveStoreModules, opts: ArchiveStoreInitOpts, signal: AbortSignal) {
+    this.chain = modules.chain;
+    this.db = modules.db;
+    this.logger = modules.logger;
+    this.metrics = modules.metrics;
+    this.opts = opts;
+    this.signal = signal;
+    this.archiveMode = opts.archiveMode;
     this.archiveBlobEpochs = opts.archiveBlobEpochs;
-    this.prevFinalized = chain.forkChoice.getFinalizedCheckpoint();
+    this.prevFinalized = this.chain.forkChoice.getFinalizedCheckpoint();
+
     this.jobQueue = new JobItemQueue<[CheckpointWithHex], void>(this.processFinalizedCheckpoint, {
-      maxLength: PROCESS_FINALIZED_CHECKPOINT_QUEUE_LEN,
+      maxLength: PROCESS_FINALIZED_CHECKPOINT_QUEUE_LENGTH,
       signal,
     });
 
+    if (opts.archiveMode === ArchiveMode.Frequency) {
+      this.statesArchiverStrategy = new FrequencyStateArchiveStrategy(
+        this.chain.regen,
+        this.db,
+        this.logger,
+        opts,
+        this.chain.bufferPool
+      );
+    } else {
+      throw new Error(`State archive strategy "${opts.archiveMode}" currently not supported.`);
+    }
+
     if (!opts.disableArchiveOnCheckpoint) {
       this.chain.emitter.on(ChainEvent.forkChoiceFinalized, this.onFinalizedCheckpoint);
       this.chain.emitter.on(ChainEvent.checkpoint, this.onCheckpoint);
 
-      signal.addEventListener(
+      this.signal.addEventListener(
         "abort",
         () => {
           this.chain.emitter.off(ChainEvent.forkChoiceFinalized, this.onFinalizedCheckpoint);
@@ -63,11 +87,73 @@ export class Archiver {
     }
   }
 
-  /** Archive latest finalized state */
+  async init(): Promise<void> {
+    if (this.opts.pruneHistory) {
+      // prune ALL stale data before starting
+      this.logger.info("Pruning historical data");
+      await callFnWhenAwait(
+        pruneHistory(
+          this.chain.config,
+          this.db,
+          this.logger,
+          this.metrics,
+          this.opts.anchorState.finalizedCheckpoint.epoch,
+          this.chain.clock.currentEpoch
+        ),
+        () => this.logger.info("Still pruning historical data, please wait..."),
+        30_000,
+        this.signal
+      );
+    }
+
+    this.historicalStateRegen = await HistoricalStateRegen.init({
+      opts: {
+        genesisTime: this.chain.clock.genesisTime,
+        dbLocation: this.opts.dbName,
+      },
+      config: this.chain.config,
+      metrics: this.metrics,
+      logger: this.logger,
+      signal: this.signal,
+    });
+  }
+
+  async close(): Promise<void> {
+    await this.historicalStateRegen?.close();
+  }
+
+  async scrapeMetrics(): Promise<string> {
+    return this.historicalStateRegen?.scrapeMetrics() ?? "";
+  }
+
+  async getHistoricalStateBySlot(
+    slot: number
+  ): Promise<{state: Uint8Array; executionOptimistic: boolean; finalized: boolean} | null> {
+    const finalizedBlock = this.chain.forkChoice.getFinalizedBlock();
+
+    if (slot >= finalizedBlock.slot) {
+      return null;
+    }
+
+    // request for finalized state using historical state regen
+    const stateSerialized = await this.historicalStateRegen?.getHistoricalState(slot);
+    if (!stateSerialized) {
+      return null;
+    }
+
+    return {state: stateSerialized, executionOptimistic: false, finalized: true};
+  }
+
+  /**
+   * Archive latest finalized state
+   * */
   async persistToDisk(): Promise<void> {
     return this.statesArchiverStrategy.archiveState(this.chain.forkChoice.getFinalizedCheckpoint());
   }
 
+  //-------------------------------------------------------------------------
+  // Event handlers
+  //-------------------------------------------------------------------------
   private onFinalizedCheckpoint = async (finalized: CheckpointWithHex): Promise<void> => {
     return this.jobQueue.push(finalized);
   };
@@ -81,7 +167,7 @@ export class Archiver {
     );
 
     this.statesArchiverStrategy.onCheckpoint(headStateRoot, this.metrics).catch((err) => {
-      this.logger.error("Error during state archive", {stateArchiveMode: this.stateArchiveMode}, err);
+      this.logger.error("Error during state archive", {archiveMode: this.archiveMode}, err);
     });
   };
 
@@ -121,7 +207,7 @@ export class Archiver {
 
       // tasks rely on extended fork choice
       const prunedBlocks = this.chain.forkChoice.prune(finalized.rootHex);
-      await this.updateBackfillRange(finalized);
+      await updateBackfillRange({chain: this.chain, db: this.db, logger: this.logger}, finalized);
 
       this.logger.verbose("Finish processing finalized checkpoint", {
         epoch: finalizedEpoch,
@@ -132,47 +218,4 @@ export class Archiver {
       this.logger.error("Error processing finalized checkpoint", {epoch: finalized.epoch}, e as Error);
     }
   };
-
-  /**
-   * Backfill sync relies on verified connected ranges (which are represented as key,value
-   * with a verified jump from a key back to value). Since the node could have progressed
-   * ahead from, we need to save the forward progress of this node as another backfill
-   * range entry, that backfill sync will use to jump back if this node is restarted
-   * for any reason.
-   * The current backfill has its own backfill entry from anchor slot to last backfilled
-   * slot. And this would create the entry from the current finalized slot to the anchor
-   * slot.
-   */
-  private updateBackfillRange = async (finalized: CheckpointWithHex): Promise<void> => {
-    try {
-      // Mark the sequence in backfill db from finalized block's slot till anchor slot as
-      // filled.
-      const finalizedBlockFC = this.chain.forkChoice.getBlockHex(finalized.rootHex);
-      if (finalizedBlockFC && finalizedBlockFC.slot > this.chain.anchorStateLatestBlockSlot) {
-        await this.db.backfilledRanges.put(finalizedBlockFC.slot, this.chain.anchorStateLatestBlockSlot);
-
-        // Clear previously marked sequence till anchorStateLatestBlockSlot, without
-        // touching backfill sync process sequence which are at
-        // <=anchorStateLatestBlockSlot i.e. clear >anchorStateLatestBlockSlot
-        // and < currentSlot
-        const filteredSeqs = await this.db.backfilledRanges.entries({
-          gt: this.chain.anchorStateLatestBlockSlot,
-          lt: finalizedBlockFC.slot,
-        });
-        this.logger.debug("updated backfilledRanges", {
-          key: finalizedBlockFC.slot,
-          value: this.chain.anchorStateLatestBlockSlot,
-        });
-        if (filteredSeqs.length > 0) {
-          await this.db.backfilledRanges.batchDelete(filteredSeqs.map((entry) => entry.key));
-          this.logger.debug(
-            `Forward Sync - cleaned up backfilledRanges between ${finalizedBlockFC.slot},${this.chain.anchorStateLatestBlockSlot}`,
-            {seqs: JSON.stringify(filteredSeqs)}
-          );
-        }
-      }
-    } catch (e) {
-      this.logger.error("Error updating backfilledRanges on finalization", {epoch: finalized.epoch}, e as Error);
-    }
-  };
 }
```

nflaig replied:
@matthewkeil genuinely curious how you reviewed this PR

nazarhussain replied:
> this is how the actual diff should have looked like if the file renaming was done separately

The renaming of files was done in separate PR. https://github.com/ChainSafe/lodestar/pull/7518

nflaig replied:
> The renaming of files was done in separate PR. https://github.com/ChainSafe/lodestar/pull/7518

Having 2 separate PRs to unstable would have been preferable

nflaig commented:
> This is part of the series of the PRs. The idea is the Archives, Historical State and Archiver have similar scope in the system, these all deals with the data which is to be archived or archived already. So we converge those all into one module called ArchiveStore , it will be responsible to storing and returning archives when needed.

@nazarhussain the **Motivation** of a change should state **why** you are doing something NOT **what** you are doing, this makes it easier to understand the rationale and goals of PRs as a reviewer

nazarhussain commented:
> > This is part of the series of the PRs. The idea is the Archives, Historical State and Archiver have similar scope in the system, these all deals with the data which is to be archived or archived already. So we converge those all into one module called ArchiveStore , it will be responsible to storing and returning archives when needed.
> 
> @nazarhussain the **Motivation** of a change should state **why** you are doing something NOT **what** you are doing, this makes it easier to understand the rationale and goals of PRs as a reviewer

The firs sentence explains why and second sentence explains what. If it's not clear to you, can explain further.

nflaig commented:
do you mean this?

> The idea is the Archives, Historical State and Archiver have similar scope in the system, these all deals with the data which is to be archived or archived already

this explains the current status quo, not why you are refactoring the code

wemeetagain commented:
:tada: This PR is included in [v1.29.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.29.0) :tada:
